NewHaven/src/ResourceTracker.cpp

```diff
@@ -0,0 +1,378 @@
+//
+// Created by hyperion on 2020-02-27.
+//
+
+#include "ResourceTracker.h"
+#include "../src/GBMap.h"
+#include "../src/Square.h"
+#include "../src/Resources.h"
+#include <map>
+#include <deque>
+#include "boost/graph/adjacency_list.hpp"
+#include "boost/graph/graph_utility.hpp"
+
+ResourceTracker::ResourceTracker(): score{new std::map<ResourceTypes, std::uint_least16_t> }{
+}
+ResourceTracker::~ResourceTracker(){
+    delete score;
+    delete DOWN;
+    delete UP;
+    delete LEFT;
+    delete RIGHT;
+    
+}
+ResourceTracker::ResourceTracker(const ResourceTracker& scores): score{ new std::map<ResourceTypes, std::uint_least16_t> }{
+    *score =*scores.score;
+}
+ResourceTracker& ResourceTracker::operator=(const ResourceTracker& scores){
+    if(&scores == this)
+        return *this;
+    *score = *scores.score;
+    return *this;
+}
+void ResourceTracker::computeScore(ResourceTrails trail) {
+    score->clear();
+    // setup -- first step is to add the elements of the main root to the total for each resources
+    deque<NodeID> queue;
+    Map map;
+    int num_vertices = boost::num_vertices(trail);
+    auto vertices = trail.vertex_set();
+    NodeID root = vertices[0];
+    // Add the starting resources from the root to the count?
+    queue.push_back(root);
+    ResourceTypes *r_resources = (*trail[root].tile).getTileContent();
+    Quad *quad = new Quad();
+            for(int i = 0; i < 4; i++){
+               ResourceTypes resource = r_resources[i];
+               addResources(resource);
+                quad->isMatching[i] = true;
+            }
+            (*quad->current_visit_count)++;
+            map.insert(Pair(root, quad));
+    // delete r_resources
+    delete r_resources;
+    while(!queue.empty()){
+        root = queue.front();
+        Quad *root_quad = map[root];
+        // now we build the queues
+        ResourceTrails::adjacency_iterator neighbourIt, neighbourEnd;
+        tie(neighbourIt, neighbourEnd) = adjacent_vertices(root, trail);
+        ResourceTypes *root_resources = (*trail[root].tile).getTileContent();
+        for (; neighbourIt != neighbourEnd; ++neighbourIt) {
+            // get ourselves our vertex
```

Comment:
You can put the tie statement in the loop to make it more understandable and local to the loop.